snuba/query/matcher.py

```diff
@@ -0,0 +1,218 @@
+        if not result:
+            return None
+
+        return result.merge(
+            MatchResult(
+                expressions={self.name: node} if isinstance(node, Expression) else {},
+                strings={self.name: node} if isinstance(node, str) else {},
+                ints={self.name: node} if isinstance(node, int) else {},
+                floats={self.name: node} if isinstance(node, float) else {},
+                dates={self.name: node} if isinstance(node, date) else {},
+                datetimes={self.name: node} if isinstance(node, datetime) else {},
+            )
+        )
+
+
+@dataclass(frozen=True)
+class AnyExpression(Pattern[Expression]):
+    """
+    Successfully matches any expression
+    """
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        return MatchResult()
+
+
+@dataclass(frozen=True)
+class AnyString(Pattern[Optional[str]]):
+    """
+    Successfully matches any string
+    """
+
+    def match(self, node: Optional[str]) -> Optional[MatchResult]:
+        return MatchResult()
+
+
+@dataclass(frozen=True)
+class String(Pattern[Optional[str]]):
+    """
+    Matches one specific string (or None).
+    """
+
+    value: Optional[str]
+
+    def match(self, node: Optional[str]) -> Optional[MatchResult]:
+        return MatchResult() if node == self.value else None
+
+
+@dataclass(frozen=True)
+class Or(Pattern[TNode], Generic[TNode]):
+    """
+    Union of multiple patterns. Matches if at least one is a valid match
+    and returns the first valid one.
+    """
+
+    patterns: Sequence[Pattern[TNode]]
+
+    def match(self, node: TNode) -> Optional[MatchResult]:
+        for p in self.patterns:
+            ret = p.match(node)
+            if ret:
+                return ret
+        return None
+
+
+@dataclass(frozen=True)
+class Column(Pattern[Expression]):
+    """
+    Matches a Column in an AST expression.
+    The column is defined by alias, name and table. For each a Pattern can be
+    provided. If one of these Patters is left None, that field will be ignored
+    when matching (equivalent to Any, but less verbose).
+    """
+
+    alias: Optional[Pattern[Optional[str]]] = None
+    column_name: Optional[Pattern[Optional[str]]] = None
+    table_name: Optional[Pattern[Optional[str]]] = None
+
+    def match(self, node: Expression) -> Optional[MatchResult]:
+        if not isinstance(node, ColumnExpr):
+            return None
+
+        result = MatchResult()
+        for pattern, value in (
+            (self.alias, node.alias),
+            (self.column_name, node.column_name),
+            (self.table_name, node.table_name),
+        ):
+            if pattern:
+                partial_result = pattern.match(value)
+                if not partial_result:
+                    return None
+                result = result.merge(partial_result)
+
+        return result
+
+
+@dataclass(frozen=True)
+class Function(Pattern[Expression]):
```

Comment:
Should this be `FunctionCall` to mirror the name in `snuba.query.expressions`?

Reply:
ok